Swift Turbo Modules (#813)
react-native-community/discussions-and-proposals: proposals/0000-swift-turbo-modules.md

---
title: Swift Turbo Modules
author:
- Oskar Kwasniewski 
- Riccardo Cipolleschi
date: 12-08-2024 
---

# RFC0000: Swift Turbo Modules

## Summary

This RFC aims to allow developers to write Turbo Modules using Swift. This will allow the usage of more modern language making maintenance of native modules easier and more accessible. 

## Motivation

The primary motivations for introducing Swift Turbo Modules are:
- Enhance developer experience for iOS developers working with React Native
- Allow the use of more modern language

> - Make barrier to entry lower

Reviewer: ```suggestion
- Lower the entry barrier to write a native turbo module
```

## Detailed design

> One of the problem why we can't adopt Swift in TurboModules  is the contamination of C++ ending up in user-space.

Reviewer: ```suggestion
One of the reasons why we can't adopt Swift in TurboModules is the contamination of C++ types ending up in user-space.
```

> One of the reasons why we can't adopt Swift in TurboModules is the contamination of C++ types ending up in user-space.

Reviewer 2: While i can understand not wanting to expose internal C++ types to user land, reality is that RN is based on C++, and if something should be exposed and used by library maintainers, then having a clear API that defines public vs internal is all that is required, the language is inconsequential. I think as a library developer, regardless of what we do, we will eventually have to debug or deal with those internals, and this should be fine. Though maybe its an unpopular opinion

Reviewer: I completely agree with you. But we would like to provide a cleaner and simple experience to library maintainers. Not every library needs to know how react native works under the hoods, no?

We also have an effort internally to try and define the public boundary better, with a clear distinction between what's public and what's private. The project is almost 10 years old in OSS, so work is not trivial and requires a bit of time, but we hope to get there soon!

### Current Situation
The interfaces we generate for an Objective-C turbomodules have this shape:
```objc
@protocol NativeMyTurboModuleSpec <RCTBridgeModule, RCTTurboModule>
@end
```

> Where the `RCTTurboModule` protocol requires the conforming object to implement a method with this signature:

Reviewer: ```suggestion
The `RCTTurboModule` protocol requires the conforming object to implement a method with this signature:
```

```objc
 - (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
       (const facebook::react::ObjCTurboModule::InitParams &)params;
```

> As you can see, the signature of this method contains two types from C++:

Reviewer: ```suggestion
The signature of this method contains two types from C++:
```

* `std::shared_ptr<facebook::react::TurboModule>`
* `facebook::react::ObjCTurboModule::InitParams`

### Solution

> The idea is to wrap the `getTurboModule` invocation in a `TurboModuleWrapper` object.

Reviewer: ```suggestion
The idea is to wrap the `getTurboModule` invocation in a `ModuleProvider` object.
```
or also

I think that Wrapper is not the right term for this pattern. we are calling a factory method to provide a new object, we are not wrapping a method call to adapt its interface or to decorate it with some other functionalities.

```objc
// In the .h file

@interface Native<UserDefinedName>Wrapper: TurboModuleWrapper
@end

// In the .mm file

@implementation Native<UserDefinedName>Wrapper

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
   (const facebook::react::ObjCTurboModule::InitParams &)params;
{
  return std::make_shared<facebook::react::Native<UserDefinedName>SpecJSI>(params);
}

@end
```

The user-defined TurboModule has already access to this file, and so the switch from `getTurboModule` to `getWrapper` doesn't require any additional includes.

Finally, we will have to update the `RCTTurboModuleManager` to take this new object into consideration. So we have to modify the provide Turbomodule with the following code
```objc
  // Step 2e: Return an exact sub-class of ObjC TurboModule
  std::shared_ptr<facebook::react::TurboModule> turboModule = nullptr;
  
  if ([module respondsToSelector:@selector(getTurboModule:)]) {
    turboModule = [module getTurboModule:params];
  } else if ([module respondsToSelector:@selector(getWrapper)]) {
    auto wrapper = [module getWrapper];
    turboModule = [wrapper getTurboModule:params];
  }
```

> Additionally we need to make sure that `React_Codegen` module is compatible with importing to Swift. I did a small test and adding few ifdefs to React_Codegen headers allows us to use Swift.

Author: Fixed

```swift

> import protocol React_Codegen.NativeSwiftTestLibrarySpec
> import protocol React_Codegen.TurboModuleWrapper
> import class React_Codegen.NativeSwiftTestLibraryWrapper

Reviewer: ```suggestion
import protocol ReactCodegen.NativeSwiftTestLibrarySpec
import protocol ReactCodegen.TurboModuleWrapper
import class ReactCodegen.NativeSwiftTestLibraryWrapper
```

Why not just importing `ReactCodegen`?

the `import protocol` and `import class` are kind of an advanced usage of the import statement in swift. Not many people use it or know what they are actually doing.

It could be that importing the whole `ReactCodegen` is not possible as it will make the build fail.

Author: Exactly, at this point importing whole module is not possible but that's something we can adjust.

- Additional complexity in codegen
- Setup Swift CI/CD to test if there are no regressions breaking swift builds

## Alternatives

- Use Objective-C for all native modules

- **Using a protocol for the wrapper.**
	The pro of this is that we don't have an empty implementation for the TurboModuleWrapper object. 
	The cons are various:
	* The `TurboModule` itself can't adopt the `TurboModuleWrapper` protocol as, otherwise, the C++ signature will come back to the public API of the I don't think this will work. 
	* We can't codegen the default implementation for the `getTurboModule` as we won't have the base class for the protocol.
	* We could use a protocol and create a companion object in the codegen which extends the protocol and it is returned by the TurboModule. This solution works, but adds a bit of ceremonies to the base implementation above.

- **Using a custom base class for each TurboModule**
	The only pro of this approach is that we can remove a few lines from the definition of every TurboModule. 
	The con of this approach are:
	* The C++ code remains in the public API of the TurboModule
	* We are creating a deeper inheritance chain which usually should be avoided.

> 	* We are asking to all the users to inherit from a different base class. This is hard to make backward compatible.

Reviewer 2: What about using the interop between Swift and C++? What is missing here right now when it comes to making modules? And how much work / breaking change would it need for it? I love this work, but going through ObjC is just going to mean every class still needs to deal with some level of ObjC nonsense, like the NSNumber stuff, and it is an additional layer between the code the library author writes and what is happening under the hood.

Author: Currently, what's stopping Swift and C++ interop is the lack of virtual functions support. So this could be viable once Apple adds this feature. Another issue is that the C++ interop doesn't support having C++ in the header files, which is what React Native is doing. Codegen generates Objective-C interface anyways but we need a way of hiding C++ types from Swift and this is where the ModuleProvider comes in.

Reviewer 2: <img width="1728" alt="Screenshot 2024-08-13 at 13 41 27" src="https://github.com/user-attachments/assets/10f85d37-6261-4732-8a5c-e22ca4ed8fc9">
Based on https://developer.apple.com/videos/play/wwdc2024/10136/?time=1307 from WWDC24, the first one should not be an issue moving forward when Swift 6 drops.
As for the second one I am not sure what the usage is, can you explain that one a bit more?

Reviewer: Hi there! 
There are several concerns here, to be honest.

What @okwasniewski explained is correct. I was in touch with Apple engineer and they mentioned that the inheritance model between Swift and C++ is too different to implement it properly, as of today. There are many cocnerns related to what a virtual class implies in C++ and a protocol in Swift. TL;DR: it will not happen that Swift class can inherit from C++.

On another side, Swift/Cxx interop is too young and unstable to be used concretely, imho.
I'm very scared that if we start introducing it, people that have custom project might incur in weird compilation issues and that maintenance cost will fall on our shoulder.
Instead, objc/swift interop is well documented and supported and, overall, more stable.

I agree that is perhaps non ideal in terms of ergonomics, but we can try to mitigate the problem Codegenerating as much as we can.
